src/immer/create-slice-with-immer.ts

```diff
@@ -0,0 +1,46 @@
+import { produce } from 'immer';
+
+// Utility type to infer the argument types of the actions
+type InferArgs<T> = T extends (...args: infer A) => void ? A : never;
+
+export type SliceActions<Value> = {
+  [actionName: string]: (...args: never[]) => (prev: Value) => void;
```

Comment:
On the other hand, https://immerjs.github.io/immer/return/ says

> It is also allowed to return arbitrarily other data from the producer function. But only if you didn't modify the draft. This can be useful to produce an entirely new state.

So the type should accept returning `Value`.

```suggestion
  [actionName: string]: (...args: never[]) => (prev: Value) => Value | void;
```

Reply:
```
If I understand correctly, ImmerActions is what is after produce is applied.
What produce takes is SliceActions, no?
```

ImmerActions is basically like `produce(SliceActions)`, it takes each action in the slice and wraps a produce function around it for direct mutability.

```
On the other hand, https://immerjs.github.io/immer/return/ says

...

So the type should accept returning Value.
```

Ah I hadn't considered that, you can return a `Value` so long as you don't modify the draft.

This example, using createSlice using a produce on each action:

```
import { produce } from 'immer';

const countSlice = createSlice<"count", CountSlice, CountActions>(
  {
    name: "count",
    value: {
      countNum: 0,
    },
    actions: {
      inc: (payload) => produce((draft) => {
        draft.countNum += payload.val;
      }),
      reset: () => produce(() => {
        // returns a new state but does not modify draft
        return { countNum: 0 };
      })
    },
  }
);
```

is equivalent to using createSliceWithImmer using mutable syntax directly:

```
const countSliceImmer = createSliceWithImmer<"count", CountSlice, CountActions>(
  {
    name: "count",
    value: {
      countNum: 0,
    },
    actions: {
      inc: (payload) => (state) => {
        state.countNum += payload.val;
      },
      reset: () => () => {
        // returns a new state but does not modify draft
        return { count: 0 }
      },
    },
  }
);
```

It's still important to not both modify the draft and return some value as that will result in an error, but this is a problem that exists in immer and libraries that use it such as Redux Toolkit, so this is something outside this jurisdiction 🤷